### PR TITLE
647 field group in core

### LINKED ISSUE #647: Field group functionality in core
**Update 2023:**

It has been decided that the field group module provides more options for field groups than we wish to include/support in core, so rather than move the entire module into core, we are planning to revamp the fieldgroup API and move certain types of field groups into core (groupings that already exist in core) and to leave others for the contrib module. The new fieldgroup API will become part of core's field + field_ui modules. 

Group types planed for core:
* Fieldset
* Details
* Div
* Vertical Tabs

Group types that should remain in contrib:
- HTML element
- HTML5
- Horizontal tabs
- Multipage
- Accordion

The Field Group contrib module would simultaneously release a new minor version (that depends on this version of Backdrop core) where those group types are removed from the module, and all remaining types are updated to use the core field group API (which will be similar to what's there already). The current group API will also be removed from the contrib module. 

Requirements for getting new fieldgroup contrib module in core:
- [ ] Integrate Fieldgroup API into field/field_ui modules
- [ ] Create group type for fieldset
- [ ] Create group type for details
- [ ] Create group type for div
- [ ] Create group type for vertical tabs

---

**Original Issue:** It would be pretty outstanding if we have field_group in core so that forms could be arranged a lot nicer with things like vertical tabs and such through the UI.

Requirements for getting fieldgroup contrib module in core:
- [x] Place the module into the core/modules directory
- [x] Move theme functions into field_group.theme.inc file
- [x] Remove all the field group help text and fieldset
- [x] Field Group Configure and Delete operations should use dropbutton
- [x] Field Group Configure form should use menu callback (move from an expanding section to a new page -- same as field settings).
- [x] Remove the `t()` function from all tests
- [x] Move display tests into `field_group.display.test`
- [x] Standardize on `Field group` for the UI and documentation (as opposed to `Fieldgroup` in some places).
- [x] Adjust very long lines to be under the 80 character limit for coding standards
- [x] Move display suite code to `ds` module: see https://github.com/backdrop-contrib/ds/issues/11  
- [x] Add description text for the `div` settings (what are "effect", and "speed"?)
- [x] Combine the options for `html_element` and `html5` and provide an upgrade path
- [ ] Replace the `new stdClass();` in `_field_group_get_html_classes()`.

User experience cleanup options:
- [x] N/A ~Remove the whole 'Fieldgroups' fieldset and help text~ (from previous iteration of module)
- [x] Remove the whole 'Fieldgroups' vertical tab (including the ability to clone from another display)

New features (options):
- Make field group containers collapsible
- Add a "clone" operation for a field group on the currently visible display
-  Instead of having a separate 'Add new group' row below 'Add new field', just have 'Group' as an option under 'Select a field type', and the different group types (e.g. fieldset, vertical tab, etc.) as 'Widgets'.
- (is this a good idea?) remove the ability to change the group type from the overview page (why?)
---

Related issues:
* [Switch to an "Add field" and "Add field group" action link](https://github.com/backdrop/backdrop-issues/issues/4898) (same as D8/D9):
* [Separate "Manage fields" and "Manage form"](https://github.com/backdrop/backdrop-issues/issues/5219) (field group would not appear on "Manage Fields")

Issue advocate: @jenlampton

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/647

Screen Shot 2022-05-19 at 6 42 32 PM

### CODE REVIEW

**Comment by herbdool** [comment]
The change to using Operations is a welcome change.

We'll need to still fix a few things and remove unused functions with this change, before it can be merged. I tested it against the contrib version and here are a few things to fix. I think I'll just paste the suggested updated functions:

Remove `field_group_format_settings_form`.

Update these functions:

```
function field_group_format_settings_label_validate($element, &$form_state) {
  $values = $form_state['values'];

  $group = $form_state['complete_form']['#group'];

  if ($group->format_type == 'fieldset' && ($values['formatter'] == 'collapsed' || $values['formatter'] == 'collapsible') && empty($values['label'])) {
    form_error($element, t('The label is required when formatter is collapsible or collapsed'));
  }
  if ($group->format_type == 'details' && empty($values['label'])) {
    form_error($element, t('The label is required'));
  }

}
```

and these validation functions:

```
/**
 * Validate the entered css class from the submitted format settings.
 * @param Array $element The validated element
 * @param Array $form_state The state of the form.
 */
function field_group_validate_css_class($element, &$form_state) {
  if (!empty($form_state['values']['instance_settings']['classes']) && !preg_match('!^[A-Za-z0-9-_ ]+$!', $form_state['values']['instance_settings']['classes'])) {
    form_error($element, t('The css class must include only letters, numbers, underscores and dashes.'));
  }
}

/**
 * Validate the entered id attribute from the submitted format settings.
 * @param Array $element The validated element
 * @param Array $form_state The state of the form.
 */
function field_group_validate_id($element, &$form_state) {
  if (!empty($form_state['values']['instance_settings']['id']) && !preg_match('!^[A-Za-z0-9-_]+$!', $form_state['values']['instance_settings']['id'])) {
    form_error($element, t('The id must include only letters, numbers, underscores and dashes.'));
  }
}
```

And the configure form and submit handler need to also handle the label and formatter, not just the instance_settings:

```
/**
 * Menu callback; present a form for configuring a group.
 */
function field_group_configure_form($form, &$form_state, $group, $view_mode = 'form') {
  $form['#group'] = $group;
  $form['#entity_type'] = $group->entity_type;
  $form['#bundle'] = $group->bundle;
  $form['#view_mode'] = $view_mode;

  $form['label'] = array(
    '#type' => 'textfield',
    '#title' => t('Field group label'),
    '#default_value' => $group->label,
    '#weight' => -5,
    '#element_validate' => array('field_group_format_settings_label_validate'),
  );

  $addition = module_invoke_all('field_group_format_settings', $group);
  $form += $addition;

  // Give fieldgroup implementations the chance to alter the description.
  if (!empty($addition['label']['#description'])) {
    $form['label']['#description'] = $addition['label']['#description'];
  }

  $form['actions'] = array(
    '#type' => 'actions'
  );
  $form['actions']['submit'] = array(
    '#type' => 'submit',
    '#value' => 'Update',
  );

  $form['actions']['cancel'] = array(
    '#type' => 'link',
    '#title' => 'Cancel',
    '#href' => _field_group_get_admin_path($group->entity_type, $group->bundle, $view_mode),
  );

  return $form;
}

/**
 * Submit handler for field_group_configure_form().
 */
function field_group_configure_form_submit($form, &$form_state) {
  $group = $form['#group'];
  $group->mode = $form['#view_mode'];
  $group->label = $form_state['values']['label'];
  $group->format_settings['instance_settings'] = $form_state['values']['instance_settings'];
  $group->format_settings['formatter'] = $form_state['values']['formatter'];

  field_group_group_save($group, FALSE);
  cache_clear_all('field_groups', 'cache_field');

  $admin_path = _field_group_get_admin_path($form['#entity_type'], $form['#bundle'], $form['#view_mode']);
  $form_state['redirect'] = $admin_path;
}
```

There may be more stuff that's expecting the old form which we can get rid of.

File: core/modules/field_group/field_group.field_ui.inc [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+      '#cell_attributes' => array('colspan' => $params->display_overview ? 1 : 2),
+    ),
+  );
+}
+
+/**
+ * Creates a form for field_group formatters.
+ * @param Object $group The FieldGroup object.
+ */
+function field_group_format_settings_form(&$group) {
```
Comment (herbdool) [inline core/modules/field_group/field_group.field_ui.inc]: I've left a long comment on the PR of suggested fixes to this file.

**Comment by jenlampton** [comment]
Re @herbdool's last comment:
* I have removed `field_group_validate_css_class()` - but now we have no class validation (which might be okay - we don't have it anywhere else in core either, but see below)
* I have removed `field_group_validate_id()` and made that form API element type `machine_name` instead.
* I have removed `field_group_format_settings_label_validate()` and made the field always required.

Since we have a lot of places where people can add CSS classes in core, that begs the following questions:
- should `field_group_validate_css_class()` become a core validation function?
- should we provide whole a form API element for CSS classes, that contains both the validation function, and the standard description text: `t('Separate class names with spaces. Example: <code>wrapper clearfix</code>')`
- should the description text become a constant, for easier re-use in contrib?

I still need to look into the other necessary updates for `field_group_configure_form()` to see if I took care of those yet...

**Comment by herbdool** [comment]
I haven't had a chance to look at this yet. Thanks @jenlampton.

I think a final PR will have to keep the commit history of the field_group contrib module. I can't recall exactly but some of your improvements here made back into the contrib, when I last updated it. https://github.com/backdrop-contrib/field_group/

File: core/modules/field_group/field_group.field_ui.inc [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+  if (!empty($format_settings['label']['#description'])) {
+    $form['label']['#description'] = $format_settings['label']['#description'];
+  }
+  
+  $form['actions'] = array(
+    '#type' => 'actions'
+  );
+  $form['actions']['submit'] = array(
+    '#type' => 'submit',
+    '#value' => 'Update',
```
Comment (klonos) [inline core/modules/field_group/field_group.field_ui.inc:741]:
```suggestion
    '#value' => t('Update'),
```

File: core/modules/field_group/field_group.field_ui.inc [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+  );
+  $form['actions']['cancel'] = array(
+    '#type' => 'link',
+    '#title' => 'Cancel',
```
Comment (klonos) [inline core/modules/field_group/field_group.field_ui.inc:745]:
```suggestion
    '#title' => t('Cancel'),
```

File: core/modules/field_group/field_group.module [hunk 4]
```diff
@@ -0,0 +1,2034 @@
...
+      ),
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_group_format_settings().
+ * 
+ * If the group has no format settings, default ones will be added.
+ *
```
Comment (klonos) [inline core/modules/field_group/field_group.module:452]:
```suggestion
 *
 * If the group has no format settings, default ones will be added.
 *
```

**Comment by herbdool** [comment]
@jenlampton for the record my suggestion was to *update* `field_group_validate_css_class`, etc, not remove. But perhaps you decided it was the better option.

**Comment by jenlampton** [comment]
>  for the record my suggestion was to update field_group_validate_css_class, etc, not remove.

Yeah, I thought it was weird that we don't have it anywhere else in core, but that we would have a module-specific version of a css class validator. 

I'd prefer that we either have a more generic css class validator that could be used anywhere, or even better, a css class form API element (with it's own validator). But I think either of those solutions can come in a follow-up issue, since they would need to be applied to all other css class elements in core, this one didn't seem particularly special.

**Comment by herbdool** [comment]
@jenlampton please recreate your PR by preserving the commit history of field_group contrib module. Much like was done with entity_reference https://github.com/backdrop/backdrop-issues/issues/1301 and other contrib modules that were merged into core. I haven't found a link yet to where @quicksketch described the steps for doing so. I'll provide it if I find it. This will be required before this module can be merged.

**Comment by herbdool** [comment]
@jenlampton this might be the link of how to merge in a git repo: https://stackoverflow.com/questions/1425892/how-do-you-merge-two-git-repositories#10548919

**Comment by jenlampton** [comment]
@herbdool  I followed these steps to preserve the history: https://gist.github.com/jenlampton/c23aa8e11d222e13060efab85d08a7e1

and I pushed a separate PR with that attempt, here:
https://github.com/backdrop/backdrop/pull/4289

I'm going to close this one.

**Pull request closed without merging**